File: AalsiTrader/Models/StraddleModels.swift
import Foundation

// MARK: - Engine Status
struct EngineStatus: Codable {
    let broker: BrokerStatus
    let market: MarketStatus
    let engine: EngineState
}

struct BrokerStatus: Codable {
    let connected: Bool
    var name: String?
    var label: String?
}

struct MarketStatus: Codable {
    let isOpen: Bool
    var istTime: String?
}

struct EngineState: Codable {
    let running: Bool
    var mode: TradingMode?
    var broker: String?
    var indexName: IndexName?
    var strategyType: StrategyType?
    var lastSpot: Double?
    var lastNiftySpot: Double?
    var lastUpdate: String?
}

enum TradingMode: String, Codable {
    case paper
    case live
}

enum IndexName: String, Codable {
    case NIFTY
    case BANKNIFTY
}

enum StrategyType: String, Codable {
    case short_straddle
    case short_strangle
    case iron_condor

    var displayName: String {
        switch self {
        case .short_straddle: return "Short Straddle"
        case .short_strangle: return "Short Strangle"
        case .iron_condor: return "Iron Condor"
        }
    }
}

// MARK: - Capital
struct StraddleCapital: Codable {
    let initialCapital: Double
    let currentCapital: Double
    let totalPnl: Double
    let maxDrawdownPct: Double
    let winRate: Double
    let totalTrades: Int
}

// MARK: - Position
struct StraddlePosition: Codable, Identifiable {
    let id: String
    var indexName: IndexName?
    var strategyType: StrategyType?
    var strategyLabel: String?
    var legs: [LegDetail]?
    var exitRules: ExitRules?
    let ceStrike: Double
    let peStrike: Double
    let ceEntryPremium: Double
    let peEntryPremium: Double
    var ceCurPremium: Double
    var peCurPremium: Double
    var ceDelta: Double?
    var peDelta: Double?
    let totalCollected: Double
    var totalCurrent: Double
    var unrealizedPnl: Double
    let niftyEntry: Double
    let entryTime: String
    let mode: TradingMode
    var broker: String?
    var ceOrderId: String?
    var peOrderId: String?
    var marginRequired: Double?
    var netMarginRequired: Double?
    var capitalUtilization: Double?
}

struct LegDetail: Codable, Identifiable {
    var id: String { "\(side)_\(strikePrice)" }
    let side: String  // CE or PE
    let action: String  // BUY or SELL
    let strikePrice: Double
    let entryPremium: Double
    var curPremium: Double
    var delta: Double?
    var instrumentId: String?
    var orderId: String?
}

struct ExitRules: Codable {
    var perLegSlPct: Double?
    let combinedSlPct: Double
    let profitTargetPct: Double
}

// MARK: - Straddle Trade
struct StraddleTrade: Codable, Identifiable {
    let id: String
    let tradeDate: String
    let strategyType: String
    var indexName: IndexName?
    let entryTime: String
    let exitTime: String
    let ceEntryPremium: Double
    let peEntryPremium: Double
    let ceExitPremium: Double
    let peExitPremium: Double
    let netPnl: Double
    var grossPnl: Double?
    let exitReason: String
    let mode: TradingMode
    var broker: String?
    var ceStrike: Double?
    var peStrike: Double?
    var niftyEntry: Double?
    var niftyExit: Double?
    var marginRequired: Double?
    var netMarginRequired: Double?
    var totalCollected: Double?
    var totalAtExit: Double?
    var lotSize: Int?
    var lots: Int?
    var legs: [LegDetail]?
}

File: AalsiTraderAndroid/app/src/main/java/com/aalsitrader/android/model/StraddleModels.kt
package com.aalsitrader.android.model

import kotlinx.serialization.Serializable

@Serializable
data class EngineStatus(
    val broker: BrokerStatusInfo = BrokerStatusInfo(),
    val market: MarketStatusInfo = MarketStatusInfo(),
    val engine: EngineState = EngineState(),
)

@Serializable
data class BrokerStatusInfo(
    val connected: Boolean = false,
    val error: String? = null,
    val name: String? = null,
    val label: String? = null,
)

@Serializable
data class MarketStatusInfo(
    val isOpen: Boolean = false,
    val istTime: String? = null,
)

@Serializable
data class EngineState(
    val running: Boolean = false,
    val mode: TradingMode? = null,
    val broker: String? = null,
    val indexName: IndexName? = null,
    val strategyType: StrategyType? = null,
    val lastSpot: Double? = null,
    val lastNiftySpot: Double? = null,
    val lastUpdate: String? = null,
)

@Serializable
enum class TradingMode { paper, live }

@Serializable
enum class IndexName { NIFTY, BANKNIFTY }

@Serializable
enum class StrategyType {
    short_straddle, short_strangle, iron_condor;

    val displayName: String
        get() = when (this) {
            short_straddle -> "Short Straddle"
            short_strangle -> "Short Strangle"
            iron_condor -> "Iron Condor"
        }
}

@Serializable
data class StraddleCapital(
    val initialCapital: Double = 0.0,
    val currentCapital: Double = 0.0,
    val totalPnl: Double = 0.0,
    val maxDrawdownPct: Double = 0.0,
    val winRate: Double = 0.0,
    val totalTrades: Int = 0,
)

@Serializable
data class StraddlePosition(
    val id: String,
    val indexName: IndexName? = null,
    val strategyType: StrategyType? = null,
    val strategyLabel: String? = null,
    val legs: List<LegDetail>? = null,
    val exitRules: ExitRules? = null,
    val ceStrike: Double = 0.0,
    val peStrike: Double = 0.0,
    val ceEntryPremium: Double = 0.0,
    val peEntryPremium: Double = 0.0,
    val ceCurPremium: Double = 0.0,
    val peCurPremium: Double = 0.0,
    val ceDelta: Double? = null,
    val peDelta: Double? = null,
    val totalCollected: Double = 0.0,
    val totalCurrent: Double = 0.0,
    val unrealizedPnl: Double = 0.0,
    val niftyEntry: Double = 0.0,
    val entryTime: String = "",
    val mode: TradingMode = TradingMode.paper,
    val broker: String? = null,
    val ceOrderId: String? = null,
    val peOrderId: String? = null,
    val marginRequired: Double? = null,
    val netMarginRequired: Double? = null,
    val capitalUtilization: Double? = null,
)

@Serializable
data class LegDetail(
    val side: String = "",
    val action: String = "",
    val strikePrice: Double = 0.0,
    val entryPremium: Double = 0.0,
    val curPremium: Double = 0.0,
    val delta: Double? = null,
    val instrumentId: String? = null,
    val orderId: String? = null,
)

@Serializable
data class ExitRules(
    val perLegSlPct: Double? = null,
    val combinedSlPct: Double = 0.0,
    val profitTargetPct: Double = 0.0,
)

@Serializable
data class StraddleTrade(
    val id: String,
    val tradeDate: String = "",
    val strategyType: String = "",
    val indexName: IndexName? = null,
    val entryTime: String = "",
    val exitTime: String = "",
    val ceEntryPremium: Double = 0.0,
    val peEntryPremium: Double = 0.0,
    val ceExitPremium: Double = 0.0,
    val peExitPremium: Double = 0.0,
    val netPnl: Double = 0.0,
    val grossPnl: Double? = null,
    val exitReason: String = "",
    val mode: TradingMode = TradingMode.paper,
    val broker: String? = null,
    val ceStrike: Double? = null,
    val peStrike: Double? = null,
    val niftyEntry: Double? = null,
    val niftyExit: Double? = null,
    val marginRequired: Double? = null,
    val netMarginRequired: Double? = null,
    val totalCollected: Double? = null,
    val totalAtExit: Double? = null,
    val lotSize: Int? = null,
    val lots: Int? = null,
    val legs: List<LegDetail>? = null,
)
